### PR TITLE
Parameterized local_elemwise_alloc_opt for GPU support

### PR DESCRIPTION
**Pull request by daemonmaker**
...uAlloc.

### CODE REVIEW

File: theano/tensor/opt.py [hunk 1]
```diff
@@ -1606,136 +1606,141 @@ def local_remove_all_assert(node):
...
+    
+        # The broadcast pattern of the ouptut must match the broadcast pattern of
+        # at least one of the inputs.
+        if not any([i.type.broadcastable ==
+            node.outputs[0].type.broadcastable for i in node.inputs]):
+            return False
+    
+        def dimshuffled_alloc(i):
+            return (isinstance(i.owner.op, DimShuffleOP) and
+                         i.owner.inputs[0].owner and \
```
Comment (abergeron) [inline theano/tensor/opt.py]: remove the `\`

File: theano/tensor/opt.py [hunk 2]
```diff
@@ -1606,136 +1606,141 @@ def local_remove_all_assert(node):
...
+                and i.owner.inputs[0].type != i.owner.outputs[0].type):
+                # when i.owner.inputs[0].type == i.owner.outputs[0].type we
+                # will remove that alloc later
+    
+                assert i.type.ndim == cmp_op.ndim
+                if (theano.config.experimental.local_alloc_elemwise_assert
+                    and not node.fgraph.shape_feature.same_shape(i, cmp_op)):
+                    assert_op = assert_(assert_op,
+                                        *[T.eq(i.shape[idx], cmp_op.shape[idx])\
+                                              for idx in xrange(i.type.ndim) \
```
Comment (abergeron) [inline theano/tensor/opt.py]: remove the `\`

**Comment by abergeron** [comment]
Apart from the small style fixes, this looks good.

**Comment by nouiz** [comment]
@abergeron the PR was updated.

**Comment by nouiz** [comment]
There is no GPU tests, so why did you merge it now? I think part of the problem is that on gpu, you probably need to register it at other time then specialize_device. You probably should register it like this in the file sanbox/cuda/opt.py:

local_gpu_elemwise_alloc = register_opt()(gof.local_optimizer([GpuElemwise])(
 local_elemwise_alloc_op(GpuElemwise, GpuAlloc, GpuDimShuffle)
))

**Comment by abergeron** [comment]
That was the CPU part.  The goal is to make it work on GPU, but the code to do this isn't in yet.

**Comment by daemonmaker** [comment]
Trying the different methods of registering the opt that you suggested has
not worked. In other words there is still a GpuAlloc in the computation
graph even though there should not be. The code can be found on my
local_alloc_elemwise2
branch <https://github.com/daemonmaker/Theano/tree/local_alloc_elemwise2> and
the test can be executed by running the following command from
Theano/theano/sandbox/cuda/tests:

THEANO_FLAGS=device=cpu nosetests
test_opt.py:Test_local_elemwise_alloc.test_remove_alloc_wo_dimshuffle

I will continue to debug it in the mean time.

Cheers,
Dustin

**Pull request merged**